src/foam/lib/csv/Outputter.java

```diff
@@ -0,0 +1,180 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package foam.lib.csv;
+
+import foam.core.*;
+import foam.dao.AbstractSink;
+import foam.lib.json.OutputterMode;
+
+import java.io.*;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.Iterator;
+import java.util.List;
+import java.util.TimeZone;
+import java.util.stream.Collectors;
+
+public class Outputter
+    extends AbstractSink
+{
+
+  protected ThreadLocal<SimpleDateFormat> sdf = new ThreadLocal<SimpleDateFormat>() {
+    @Override
+    protected SimpleDateFormat initialValue() {
+      SimpleDateFormat df = new SimpleDateFormat("YYYY-MM-dd'T'HH:mm:ss.S'Z'");
+      df.setTimeZone(TimeZone.getTimeZone("UTC"));
+      return df;
+    }
+  };
+
+  protected StringWriter stringWriter_ = null;
+  protected PrintWriter writer_;
+  protected OutputterMode mode_;
+  protected boolean outputHeaders_;
+  protected boolean isHeadersOutput_ = false;
+  
+  public Outputter() {
+    this(OutputterMode.FULL);
+  }
+
+  public Outputter(OutputterMode mode) {
+    this((PrintWriter) null, mode, true);
+  }
+
+  public Outputter(OutputterMode mode, boolean outputHeaders) {
+    this((PrintWriter) null, mode, outputHeaders);
+  }
+
+  public Outputter(File file, OutputterMode mode, boolean outputHeaders) throws FileNotFoundException {
+    this(new PrintWriter(file), mode, outputHeaders);
+  }
+
+  public Outputter(PrintWriter writer, OutputterMode mode, boolean outputHeaders) {
+    if ( writer == null ) {
+      stringWriter_ = new StringWriter();
+      writer = new PrintWriter(stringWriter_);
+    }
+
+    this.mode_ = mode;
+    this.writer_ = writer;
+    this.outputHeaders_ = outputHeaders;
+  }
+
+  public String stringify(FObject obj) {
+    if ( stringWriter_ == null ) {
+      stringWriter_ = new StringWriter();
+      writer_ = new PrintWriter(stringWriter_);
+    }
+
+    stringWriter_.getBuffer().setLength(0);
+    if ( outputHeaders_ )
+      outputHeaders(obj);
+    outputFObject(obj);
+    return this.toString();
+  }
+
+  /**
+   * Gets a filtered list of properties. Removes network and storage transient variables
+   * if necessary, removes unsupported types and removes null values / empty strings
+   * @param obj the object to get the property list from
+   * @return the filtered list of properties
+   */
+  public List<PropertyInfo> getFilteredPropertyInfoList(FObject obj) {
+    List<PropertyInfo> props = obj.getClassInfo().getAxiomsByClass(PropertyInfo.class);
```

Comment:
Cache this value so you don't need to build collection every time.

Reply:
That would prevent reusing an instance of this class to output different classes of FObjects.